### PR TITLE
Remove use_c10_dispatcher option

### PR DESCRIPTION
**Pull request by wenleix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54969 Remove use_c10_dispatcher option**

With all use cases to hacky wrapper removed, all kernels will be
dispatched with c10 full dispatcher.

Differential Revision: [D27436596](https://our.internmc.facebook.com/intern/diff/D27436596/)

### CODE REVIEW

File: aten/src/ATen/core/op_registration/adaption.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * We may want make the kernel handle these parameters directly without going
+ * through the creation of a TensorOptions value.
+ */
+
+namespace c10 {
+namespace impl {
+
+inline c10::optional<MemoryFormat>
+check_tensor_options_and_extract_memory_format(
```
Comment (wenleix) [inline aten/src/ATen/core/op_registration/adaption.h:42]: `check_tensor_options_and_extract_memory_format` is still required. But not sure how to rename `hacky_wrapper_for_legacy_signatures.h`. Use `adaption.h` in this PR but let me know if there is any better suggestions. cc @smessmer

Comment (ezyang) [reply aten/src/ATen/core/op_registration/adaption.h:42]: `check_tensor_options_and_extract_memory_format` is actually a good example of another thing we need to get rid of; it's due to MemoryFormat being a separate argument in some JIT schema but also being specified in TensorOptions.

**Review by ezyang — APPROVED** [review]
wooooo

File: tools/codegen/local.py [hunk 2]
```diff
@@ -17,24 +15,10 @@
 # DON'T add a new entry here.
 
 class Locals(threading.local):
-    use_c10_dispatcher: Optional[UseC10Dispatcher] = None
-_locals = Locals()
```
Comment (smessmer) [inline tools/codegen/local.py:21]: This looks like we can remove the whole Locals class, cc @ezyang did you want to use it for something else?

**Review by smessmer — APPROVED** [review]
Yay :) Stoked to finally get rid of hacky_wrapper

File: tools/codegen/local.py (hunk 2 above)
Comment (wenleix) [reply tools/codegen/local.py:21]: Discussed this offline with @ezyang . Supporting dynamic scoping during codegen seems a useful feature, plus is not very much code.

**Pull request closed without merging**